File: code/nest/lab1/l1e4_thread_limit_kernel.py
import re

def false_fork_fail(line):
    a = re.findall('Forking (.*) failed!\n', line)
    if (len(a) > 0):
        return a[0]
    else:
        return None

def false_fork_success(line):
    a = re.findall('Forking (.*) succeeded, but over limit\n', line)
    if (len(a) > 0):
        return a[0]
    else:
        return None

def check(lines):
    for line_num, line in enumerate(lines):
        thread_name = false_fork_fail(line)
        if thread_name is not None:
            error_message = 'Error in line {}: Forking {} failed.\n'\
                .format(line_num+1, thread_name)
            return False, error_message
        thread_name = false_fork_success(line)
        if thread_name is not None:
            error_message = 'Error in line {}: \
                Forking {} succeeded, but over thread limit\
                (thread \"main\" should also be counted).\n'\
                .format(line_num+1, thread_name)
            return False, error_message
    return True, None

File: code/nest/lab1/l1e4_thread_limit_kernel.cc
#include <cstring>
#include "system.h"

#ifndef THREAD_NUM_LIMIT
#define THREAD_NUM_LIMIT 128
#endif

void stub(void *p)
{
    int which = (int)p;
    DEBUG('e', "Thread %d runs successfully\n", which);
}

int Nest(void *arg) {
    DEBUG('e', "Entering Nest()\n");
    char* threadNames[THREAD_NUM_LIMIT+1];
    threadNames[0] = "main";

    for (int i = 1; i < THREAD_NUM_LIMIT; i++) 
    {
        threadNames[i] = new char[10];
        sprintf(threadNames[i], "thread%d", i);
        Thread *t = new Thread(threadNames[i]);
        int ret = t->Fork(stub, (void*)i);
        if (ret != 0) {
            DEBUG('e', "Forking %s failed!\n", threadNames[i]);
            return -1;
        } else {
            DEBUG('e', "Forking %s succeeded.\n", threadNames[i]);
        }
    }
    threadNames[THREAD_NUM_LIMIT] = new char[10];
    sprintf(threadNames[THREAD_NUM_LIMIT], "thread%d", THREAD_NUM_LIMIT);
    Thread *t = new Thread(threadNames[THREAD_NUM_LIMIT]);
    int ret = t->Fork(stub, (void*)THREAD_NUM_LIMIT);
    if (ret == 0) {
        DEBUG('e', "Forking %s succeeded, but over limit\n", threadNames[THREAD_NUM_LIMIT]);
        return -1;
    }
    return 0;
}
